### PR TITLE
Add basic auth support for HTTPS tunnel

### PR DESCRIPTION
**Pull request by moonchen**
-Accept username and password from command line
-Authorize requests based on username and password.
-The above only applies to the CONNECT tunnel over TLS

### CODE REVIEW

File: server/http_tunnel.go [hunk 1]
```diff
@@ -31,22 +32,36 @@ func StartHTTPSTunnel(addr, externalProxy, certFile, KeyFile string) {
 		}}
 		proxy.ConnectDial = proxy.NewConnectDialToProxy(externalProxy)
 	}
+
+	proxy.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
```
Comment (zhucebuliaolongchuan) [inline server/http_tunnel.go:36]: why did you need to add this? If this is just for debugging purposes, would that be okay to do this under the next func()? `proxy.OnRequest().DoFunc()`?

I could see there would be duplicate debug info when there is a connect request coming in. Some thing like:
```
$ HTTP Tunnel: got CONNECT request $HOST
$ HTTP Tunnel: got request $URL
```

File: server/http_tunnel.go [hunk 2]
```diff
@@ -31,22 +32,36 @@ func StartHTTPSTunnel(addr, externalProxy, certFile, KeyFile string) {
...
+
 	proxy.OnRequest().DoFunc(
 		func(r *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
 			// Leave the request untouched.  Log a message for debugging purposes.
 			log.Printf("HTTP Tunnel: got request %s\n", r.URL)
 			return r, nil
 		})
 
-	if certFile == "" || KeyFile == "" {
+	if username != "" || password != "" {
```
Comment (zhucebuliaolongchuan) [inline server/http_tunnel.go:48]: I am kind of curious here. Is it okay to let one of this to be empty? If both of them are not empty, and then we are going to perform the basic authentication?

**Review by zhucebuliaolongchuan — CHANGES_REQUESTED** [review]

File: server/http_tunnel.go (hunk 2 above)
Comment (moonchen) [reply server/http_tunnel.go:48]: I'm using empty string to mean that no username or password is set.  In command line parsing, we're not giving the user a way to set empty string to be the username or password.  Therefore, by the time we get here, if both are empty, that means neither command line options were specified.

File: cmd/edge-proxy/main.go [hunk 3]
```diff
@@ -68,6 +70,8 @@ func main() {
 	flag.StringVar(&httpsTunnelAddr, "https-tunnel-listen", "", "Listen address for HTTPS (CONNECT) tunnel server over TLS.  Both tunnels can be served at the same time.")
 	flag.StringVar(&httpsTunnelTLSCert, "https-tunnel-tls-cert", "", "For the HTTPS tunnel, specify file name and path to the TLS certificate /path/file.crt")
 	flag.StringVar(&httpsTunnelTLSKey, "https-tunnel-tls-key", "", "For the HTTPS tunnel, specify file name and path to the TLS key /path/file.key")
+	flag.StringVar(&httpsTunnelUsername, "https-tunnel-username", "", "For the HTTPS tunnel only, require this username for basic authentication")
+	flag.StringVar(&httpsTunnelPassword, "https-tunnel-password", "", "For the HTTPS tunnel only, require this password for basic authentication")
 	flag.Parse()
```
Comment (ryannowarm) [inline cmd/edge-proxy/main.go:75]: It's tough for me to tell if these username/password options are for the server-side or client side.  In order to make these command line options clearer, what do you think about naming all the server options: `http[s]-proxy-server-*` and all of the client options `http[s]-proxy-client-*` ?

...alternatively `connect-proxy-server-*` and `connect-proxy-client-*` ?

File: server/http_tunnel.go (hunk 1 above)
Comment (moonchen) [reply server/http_tunnel.go:36]: Connect and other requests go through different code paths.  In my testing, I never got a request that generated both outputs.  For more info, see this part of the README: https://github.com/elazarl/goproxy/blob/master/README.md#type-of-handlers-for-manipulating-connectreqresp-behavior .

File: cmd/edge-proxy/main.go (hunk 3 above)
Comment (moonchen) [reply cmd/edge-proxy/main.go:75]: Does the help text not make it clear?  We decided against using "proxy" because that word is overloaded in edge-proxy.  Connect proxy is clear, but since we can serve both plain and TLS at the same time, that gets quite verbose, e.g. `connect-proxy-over-tls-server-username`

File: server/http_tunnel.go (hunk 2 above)
Comment (moonchen) [reply server/http_tunnel.go:48]: It is possible to let one of them be empty.

File: cmd/edge-proxy/main.go (hunk 3 above)
Comment (ryannowarm) [reply cmd/edge-proxy/main.go:75]: That's right, I remember the whole overloaded proxy term discussion.  How about just adding more verbose description?  So instead of
> For the HTTPS tunnel only, require this username for basic authentication

...something like:
> For the HTTPS CONNECT proxy server, require this username for basic authentication

Comment (costanic) [reply cmd/edge-proxy/main.go:75]: Do the username/password options conflict with cert/key options?  What happens if the user  tries to specify  both?

File: server/http_tunnel.go (hunk 2 above)
Comment (costanic) [reply server/http_tunnel.go:48]: It's possible for the password to be empty, but it doesn't make sense for the username to be empty if a password was specified.  How is that case handled?

File: server/http_tunnel.go [hunk 4]
```diff
@@ -31,22 +32,36 @@ func StartHTTPSTunnel(addr, externalProxy, certFile, KeyFile string) {
...
+		auth.ProxyBasic(proxy, "tunnel", func(user, passwd string) bool {
+			authorized := (user == username && passwd == password)
+			log.Printf("HTTP Tunnel: user=%s, password=%s, authorized=%t\n", user, passwd, authorized)
```
Comment (costanic) [inline server/http_tunnel.go:51]: we're not really logging usernames and passwords right?  are these cleartext?

**Review by costanic — CHANGES_REQUESTED** [review]

File: cmd/edge-proxy/main.go (hunk 3 above)
Comment (moonchen) [reply cmd/edge-proxy/main.go:75]: There's a possible interaction between these options: the username/password only applies to the HTTPS tunnel.  The HTTPS tunnel requires tls cert and key.  I will add validation to make sure that cert and key is specified when user/pass are specified.

File: server/http_tunnel.go (hunk 2 above)
Comment (moonchen) [reply server/http_tunnel.go:48]: It's possible to form a valid basic auth header with an empty username or an empty password, or both.  But since it's confusing, I'll require both to be set on the command line.

Comment (costanic) [reply server/http_tunnel.go:48]: Nah, that's ok.  If it's actually a valid scenario where a proxy is configured to have an empty username but a nonempty password, then we should allow it.

File: server/http_tunnel.go (hunk 1 above)
Comment (zhucebuliaolongchuan) [reply server/http_tunnel.go:36]: Sounds good.

**Review by zhucebuliaolongchuan — APPROVED** [review]

File: server/http_tunnel.go (hunk 4 above)
Comment (moonchen) [reply server/http_tunnel.go:51]: We're logging the username and passwords that people send us.  It is cleartext once it gets to us.

File: cmd/edge-proxy/main.go (hunk 3 above)
Comment (moonchen) [reply cmd/edge-proxy/main.go:75]: Validation added.

File: server/http_tunnel.go (hunk 4 above)
Comment (moonchen) [reply server/http_tunnel.go:51]: They're encrypted during transit by TLS.

Comment (costanic) [reply server/http_tunnel.go:51]: we shouldn't log usernames and passwords.

Comment (moonchen) [reply server/http_tunnel.go:51]: Why not?

Comment (costanic) [reply server/http_tunnel.go:51]: Because usernames and passwords are sensitive customer data and we don't want to be responsible for leaking it.

Beyond writing these to a log file, why are we manually comparing usernames and passwords.  Does the library not handle that?

Comment (costanic) [reply server/http_tunnel.go:51]: @moonchen Here's an example of [a very high profile company that had to issue a public statement](https://blog.twitter.com/official/en_us/topics/company/2018/keeping-your-account-secure.html) when it was discovered they were writing customer passwords to a log file.  The statement goes on to request that all users reset their passwords.  We don't want to be responsible for a leak like this.

> Due to a bug, passwords were written to an internal log before completing the hashing process. We found this error ourselves, removed the passwords, and are implementing plans to prevent this bug from happening again.

Comment (moonchen) [reply server/http_tunnel.go]: The correct username and password are available on the command line, but I see the possibility that the user may accidentally use a incorrect but sensitive password in an unsuccessful attempt.  I'll remove the info from the log.

Comment (moonchen) [reply server/http_tunnel.go]: > Because usernames and passwords are sensitive customer data and we don't want to be responsible for leaking it.
> 
> Beyond writing these to a log file, why are we manually comparing usernames and passwords. Does the library not handle that?

The library doesn't handle the comparison for us.

**Comment by costanic** [comment]
Thanks for removing the user/pass log.  Can you comment on why we manually compare the username & password?  Can the library handle that for us?

**Comment by costanic** [comment]
> Thanks for removing the user/pass log. Can you comment on why we manually compare the username & password? Can the library handle that for us?

Oh I see the reply now.  Which library are we using?  I thought it was goproxy?  https://github.com/coodajingang/goproxy/blob/bb933cfb6e58/utils/structs.go#L252

**Comment by moonchen** [comment]
> > Thanks for removing the user/pass log. Can you comment on why we manually compare the username & password? Can the library handle that for us?
> 
> Oh I see the reply now. Which library are we using? I thought it was goproxy? https://github.com/coodajingang/goproxy/blob/bb933cfb6e58/utils/structs.go#L252

We're using a different library, also named goproxy.  The relevant code is here: https://github.com/elazarl/goproxy/blob/a92cc753f88eb1d5f3ca49bd91da71fe815537ca/ext/auth/basic.go#L33

File: server/http_tunnel.go (hunk 4 above)
Comment (ryannowarm) [reply server/http_tunnel.go]: Just my 2 cents on password logging.  Even if the password is viewable because it's in the clear as a command line arg, doesn't mean that it should be clear in the log.  For example, there's nothing stopping our users from forwarding these logs via journald off to some log aggregator(ex: Pelion Cloud).  In that case, we've leaked their password out of the box.  Users wouldn't expect this because the general rule is that you never log secrets of any kind including passwords.

**Comment by costanic** [comment]
> > > Thanks for removing the user/pass log. Can you comment on why we manually compare the username & password? Can the library handle that for us?
> > 
> > 
> > Oh I see the reply now. Which library are we using? I thought it was goproxy? https://github.com/coodajingang/goproxy/blob/bb933cfb6e58/utils/structs.go#L252
> 
> We're using a different library, also named goproxy. The relevant code is here: https://github.com/elazarl/goproxy/blob/a92cc753f88eb1d5f3ca49bd91da71fe815537ca/ext/auth/basic.go#L33

Well that's confusing.

**Review by costanic — APPROVED** [review]
